File: csharp/AdventOfCode.2021/Challenges/Challenge06.cs
using AdventOfCode.Core;
using AdventOfCode.Core.IO;
using AdventOfCode.Lib;

namespace AdventOfCode2021.Challenges;

[Challenge(6)]
public class Challenge06(IInputReader InputReader)
{
    [Part1]
    public async Task<string> Part1Async()
    {
        var data = await InputReader.ReadLineAsync<int>(6, ',').ToListAsync();

        var groups = new ulong[9];
        foreach (var i in data)
            groups[i]++;

        return CalculateFishies(80, groups).ToString();
    }

    [Part2]
    public async Task<string> Part2Async()
    {
        var data = await InputReader.ReadLineAsync<int>(6, ',').ToListAsync();

        var groups = new ulong[9];
        foreach (var i in data)
            groups[i]++;

        return CalculateFishies(256, groups).ToString();
    }

    private ulong CalculateFishies(int totalDays, ulong[] groups) => Breed(0, totalDays, groups).Sum(x => x);

    private static ulong[] Breed(int dayFrom, int totalDays, ulong[] groups)
    {
        if (dayFrom == totalDays)
            return groups;

        return Breed(dayFrom + 1, totalDays,
            new[]
            {
                groups[1], groups[2], groups[3], groups[4], groups[5], groups[6], groups[0] + groups[7], groups[8],
                groups[0]
            });
    }
}

File: fsharp/AdventOfCode.2021/AdventOfCode.2021/Challenges/Challenge06.fs
module Challenges.Challenge06

open Utils.IO

let groupFish (data:int64 array) =
    let grouped = data
                |> Array.groupBy (fun x -> x)
                |> Array.map (fun (a,b) -> (int a,b))
                |> Map.ofArray

    [|for i in 0..8 -> int64 i|]
    |> Array.mapi (
        fun i _ ->
            match Map.tryFind i grouped with
            | Some v -> int64 (Array.length v)
            | None -> int64 0
        )

let rec breed dayFrom totalDays groups =
    match dayFrom with
    | d when d = totalDays -> groups
    | d -> breed (d+1) totalDays [| groups.[1]; groups.[2]; groups.[3]; groups.[4]; groups.[5]; groups.[6] ; groups.[7] + groups.[0]; groups.[8]; groups.[0] |]

let calculateFishies totalDays groups =
    breed 0 totalDays groups
    |> Array.sum

let setup = fun () -> readLine<int64> 6 ','

let part1 input =
    input
    |> groupFish
    |> calculateFishies 80
    |> string

let part2 input =
    input
    |> groupFish
    |> calculateFishies 256
    |> string